src/Plugin/Commerce/PaymentGateway/ExpressCheckout.php

```diff
@@ -0,0 +1,570 @@
+      '#required' => TRUE,
+    ];
+    $form['api_password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('API Password'),
+      '#default_value' => $this->configuration['api_password'],
+      '#required' => TRUE,
+    ];
+    $form['signature'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Signature'),
+      '#default_value' => $this->configuration['signature'],
+      '#required' => TRUE,
+    ];
+
+    $form['solution_type'] = [
+      '#type' => 'radios',
+      '#title' => t('Type of checkout flow'),
+      '#description' => t('Express Checkout Account Optional (ECAO) where PayPal accounts are not required for payment may not be available in all markets.'),
+      '#options' => [
+        'Mark' => t('Require a PayPal account (this is the standard configuration).'),
+        'SoleLogin' => t('Allow PayPal AND credit card payments, defaulting to the PayPal form.'),
+        'SoleBilling' => t('Allow PayPal AND credit card payments, defaulting to the credit card form.'),
+      ],
+      '#default_value' => $this->configuration['solution_type'],
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    parent::submitConfigurationForm($form, $form_state);
+    if (!$form_state->getErrors()) {
+      $values = $form_state->getValue($form['#parents']);
+      $this->configuration['api_username'] = $values['api_username'];
+      $this->configuration['api_password'] = $values['api_password'];
+      $this->configuration['signature'] = $values['signature'];
+      $this->configuration['solution_type'] = $values['solution_type'];
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function onReturn(OrderInterface $order, Request $request) {
+    // GetExpressCheckoutDetails API Operation (NVP).
+    // Shows information about an Express Checkout transaction.
+    $paypal_response = $this->apiGetExpressCheckoutDetails($order);
+
+    // If the request failed, exit now with a failure message.
+    if ($paypal_response['ACK'] == 'Failure') {
+      return FALSE;
+    }
+
+    // Set the Payer ID used to finalize payment.
+    $order_express_checkout_data = $order->getData('paypal_express_checkout');
+    $order_express_checkout_data['payerid'] = $paypal_response['PAYERID'];
+    $order->setData('paypal_express_checkout', $order_express_checkout_data);
+
+    // If the user is anonymous, add their PayPal e-mail to the order.
+    if (empty($order->mail)) {
+      $order->setEmail($paypal_response['EMAIL']);
+    }
+    $order->save();
+
+    // DoExpressCheckoutPayment API Operation (NVP).
+    // Completes an Express Checkout transaction.
+    $paypal_response = $this->apiDoExpressCheckoutDetails($order);
+
+    // Nothing to do for failures for now - no payment saved.
+    // ToDo - more about the failures.
+    if ($paypal_response['PAYMENTINFO_0_PAYMENTSTATUS'] == 'Failed') {
+      return FALSE;
```

Comment:
Per the `OffsitePaymentGatewayInterface` documentation:

```
   * @throws \Drupal\commerce_payment\Exception\PaymentGatewayException
   *   Thrown when the request is invalid or the payment failed.
```

We need to throw an exception and not just return `FALSE`.